File: src/main/java/thedarkcolour/futuremc/entity/trident/EntityTrident.java
package thedarkcolour.futuremc.entity.trident;

import net.minecraft.dispenser.IPosition;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.MoverType;
import net.minecraft.entity.effect.EntityLightningBolt;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.play.server.SPacketChangeGameState;
import net.minecraft.util.DamageSource;
import net.minecraft.util.EntityDamageSourceIndirect;
import net.minecraft.util.EntitySelectors;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import thedarkcolour.futuremc.FutureMC;
import thedarkcolour.futuremc.enchantment.EnchantHelper;
import thedarkcolour.futuremc.enchantment.Enchantments;
import thedarkcolour.futuremc.sound.Sounds;

import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

public class EntityTrident extends EntityModArrow {
    private static final Predicate<Entity> ARROW_TARGETS = entity -> EntitySelectors.NOT_SPECTATING.test(entity) && EntitySelectors.IS_ALIVE.test(entity) && entity.canBeCollidedWith();
    private ItemStack thrownStack;
    private boolean hasChanneled;
    private int ticksInGround = 0;
    private boolean isReturning;

    public EntityTrident(World world) {
        super(world);
    }

    public EntityTrident(World world, IPosition pos, ItemStack stack) {
        super(world, pos.getX(), pos.getY(), pos.getZ());
        this.thrownStack = stack;
        this.pickupStatus = PickupStatus.ALLOWED;
    }

    public EntityTrident(World world, EntityLivingBase shooter, ItemStack stack) {
        super(world, shooter);
        this.thrownStack = stack;
    }

    @Override
    protected ItemStack getArrowStack() {
        return this.thrownStack.copy();
    }

    @Override
    protected Entity findEntityOnPath(Vec3d start, Vec3d end) {
        if (world.isRemote) {
            return null;
        }
        Entity entity = null;
        List<Entity> list = this.world.getEntitiesInAABBexcluding(this, this.getEntityBoundingBox().expand(this.motionX, this.motionY, this.motionZ).grow(1.0D), ARROW_TARGETS::test);
        double d0 = 0.0D;

        for (Entity entity1 : list) {
            if (entity1 != this.shootingEntity) {
                AxisAlignedBB axisalignedbb = entity1.getEntityBoundingBox().grow(0.30000001192092896D);
                RayTraceResult raytraceresult = axisalignedbb.calculateIntercept(start, end);

                if (raytraceresult != null) {
                    double d1 = start.squareDistanceTo(raytraceresult.hitVec);

                    if (d1 < d0 || d0 == 0.0D) {
                        entity = entity1;
                        d0 = d1;
                    }
                }
            }
        }
        return entity;
    }

    @Override
    public void writeEntityToNBT(NBTTagCompound compound) {
        super.writeEntityToNBT(compound);
        compound.setTag("ItemTrident", this.thrownStack.writeToNBT(new NBTTagCompound()));
    }

    @Override
    public void readEntityFromNBT(NBTTagCompound compound) {
        super.readEntityFromNBT(compound);
        if (compound.hasKey("ItemTrident")) {
            this.thrownStack = read(compound.getCompoundTag("ItemTrident"));
        }
    }

    @Override
    protected void onHit(RayTraceResult result) {
        Entity entity = result.entityHit;

        if (entity != null) {
            DamageSource source;

            if (!(this.shootingEntity instanceof EntityPlayer)) {
                source = causeTridentDamage(this, this);
            } else {
                source = causeTridentDamage(this, shootingEntity);
            }

            if (entity.attackEntityFrom(source, getDamageForTrident(entity))) {
                if (entity instanceof EntityLivingBase) {
                    EntityLivingBase entityLivingBase = (EntityLivingBase) entity;

                    if(this.shootingEntity instanceof EntityLivingBase) {
                        EnchantmentHelper.applyThornEnchantments(entityLivingBase, this.shootingEntity);
                        EnchantmentHelper.applyArthropodEnchantments(entityLivingBase, this.shootingEntity);
                    }

                    this.arrowHit(entityLivingBase);

                    if(this.shootingEntity != null && entityLivingBase != this.shootingEntity && entityLivingBase instanceof EntityPlayer && this.shootingEntity instanceof EntityPlayer) {
                        ((EntityPlayerMP)this.shootingEntity).connection.sendPacket(new SPacketChangeGameState(6, 0.0F));
                    }

                    this.playSound(Sounds.TRIDENT_PIERCE, 1.0F, 1.0F);

                    this.motionX *= -0.009999999776482582D;
                    this.motionY *= -0.10000000149011612D;
                    this.motionZ *= -0.009999999776482582D;
                    this.rotationYaw += 180.0F;
                    this.prevRotationYaw += 180.0F;

                    if(!this.hasChanneled && thrownStack != null && !thrownStack.isEmpty()) { // Enchantment handling
                        Map enchants = EnchantmentHelper.getEnchantments(thrownStack);

                        if(!enchants.isEmpty()) {
                            if (enchants.get(Enchantments.CONDUCTIVIDAD) != null) {
                                if(world.isThundering()) {
                                    shootingEntity.world.addWeatherEffect(new EntityLightningBolt(shootingEntity.world, this.posX, this.posY, this.posZ, false));
                                    this.playSound(Sounds.TRIDENT_CONDUCTIVIDAD, 5.0F, 1.0F);
                                    this.hasChanneled = true;
                                }
                            }
                        }
                    }
                }
            } else {
                if (!this.world.isRemote && this.motionX * this.motionX + this.motionY * this.motionY + this.motionZ * this.motionZ < 0.0010000000474974513D) {
                    if (this.pickupStatus == EntityModArrow.PickupStatus.ALLOWED) {
                        this.entityDropItem(this.getArrowStack(), 0.1F);
                    }
                    this.setDead();
                }
            }
        } else {
            this.motionX = (float)(result.hitVec.x - this.posX);
            this.motionY = (float)(result.hitVec.y - this.posY);
            this.motionZ = (float)(result.hitVec.z - this.posZ);
            float f2 = MathHelper.sqrt(this.motionX * this.motionX + this.motionY * this.motionY + this.motionZ * this.motionZ);
            this.posX -= this.motionX / (double)f2 * 0.05000000074505806D;
            this.posY -= this.motionY / (double)f2 * 0.05000000074505806D;
            this.posZ -= this.motionZ / (double)f2 * 0.05000000074505806D;
            this.inGround = true;
            this.arrowShake = 7;

            this.playSound(Sounds.TRIDENT_IMPACT, 1.0F, 1.0F);

            if(!this.hasChanneled && EnchantHelper.hasChanneling(thrownStack)) { // Enchantment handling
                Entity living = shootingEntity == null ? this : shootingEntity;
                living.world.addWeatherEffect(new EntityLightningBolt(living.world, this.posX, this.posY, this.posZ, false));
                this.playSound(Sounds.TRIDENT_CONDUCTIVIDAD, 5.0F, 1.0F);
                this.hasChanneled = true;
            }
        }
    }

    @Override
    public void onUpdate() {
        if(this.inGround) {
            this.ticksInGround++;
        }

        if(shootingEntity instanceof EntityPlayer) {
            int loyalty = EnchantHelper.getLoyalty(thrownStack);

            if(0 < loyalty && loyalty < 4 && !((EntityPlayer) shootingEntity).isSpectator() && !(shootingEntity.isDead) && ticksInGround > 100 - (loyalty * 20)) {
                this.noClip = true;

                if(!isReturning) {
                    this.playSound(Sounds.TRIDENT_LOYALTY, 1.0F, 1.0F);
                    this.isReturning = true;
                }
                if(this.ticksInGround >= (100 - (loyalty * 20))){
                    double d0 = 0.02D * (double) loyalty;
                    Vec3d vec3d = new Vec3d(shootingEntity.posX - this.posX, shootingEntity.posY
                            + (double) shootingEntity.getEyeHeight() - this.posY, shootingEntity.posZ - this.posZ);
                    this.motionX += vec3d.x * d0 - this.motionX * 0.05D;
                    this.motionY += vec3d.y * d0 - this.motionY * 0.05D;
                    this.motionZ += vec3d.z * d0 - this.motionZ * 0.05D;

                    this.move(MoverType.SELF, this.motionX * 0.05D, this.motionY * 0.05D, this.motionZ * 0.05D);
                }
            }
        }

        if(this.world.getClosestPlayerToEntity(this, 1.5D) == this.shootingEntity && isReturning) {
            if(this.pickupStatus == PickupStatus.ALLOWED) {
                dropTridentStack();
            }
            this.setDead();
        }

        super.onUpdate();
    }

    @Override
    public boolean getIsCritical() {
        return false;
    }

    @Override
    public boolean isBurning() {
        return false;
    }

    @Override
    public void setFire(int seconds) {
        // Fireproof
    }

    private static ItemStack read(NBTTagCompound compound) {
        try {
            return new ItemStack(compound);
            // Not sure if i need a try-catch, im scared to delete this
        } catch (RuntimeException runtimeexception) {
            FutureMC.logger.debug("Tried to load invalid item: {}", compound, runtimeexception);
            return ItemStack.EMPTY;
        }
    }

    private float getDamageForTrident(Entity target) {
        float level = EnchantHelper.getImpaling(this.thrownStack);
        float damage = 8F;

        if(level > 0 && target.isWet()) {
            damage += (level * 1.25);
        }
        return damage;
    }

    private void dropTridentStack() {
        if(thrownStack != null && !thrownStack.isEmpty()) {
            EntityItem entityitem = new EntityItem(this.world, this.posX, this.posY + (double) (float) 0.1, this.posZ, thrownStack);
            entityitem.setNoPickupDelay();

            if (captureDrops) {
                this.capturedDrops.add(entityitem);
            }
            else {
                this.world.spawnEntity(entityitem);
            }
        }
    }

    private static DamageSource causeTridentDamage(Entity trident, Entity indirectEntityIn) {
        return (new EntityDamageSourceIndirect("trident", trident, indirectEntityIn)).setProjectile();
    }

    @Override
    protected float getWaterDrag() {
        return 0.99F;
    }
}

File: src/main/java/thedarkcolour/futuremc/entity/trident/EntityTrident.kt
package thedarkcolour.futuremc.entity.trident

import net.minecraft.dispenser.IPosition
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.MoverType
import net.minecraft.entity.effect.EntityLightningBolt
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.play.server.SPacketChangeGameState
import net.minecraft.util.DamageSource
import net.minecraft.util.EntityDamageSourceIndirect
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.RayTraceResult
import net.minecraft.util.math.Vec3d
import net.minecraft.world.World
import thedarkcolour.futuremc.enchantment.EnchantHelper
import thedarkcolour.futuremc.init.Sounds

class EntityTrident : EntityModArrow {
    private var thrownStack: ItemStack = ItemStack.EMPTY
    private var hasChanneled: Boolean = false
    private var ticksInGround = 0
    private var isReturning: Boolean = false

    override val waterDrag: Float
        get() = 0.99f

    @Suppress("unused")
    constructor(world: World) : super(world)

    constructor(world: World, pos: IPosition, stack: ItemStack) : super(world, pos.x, pos.y, pos.z) {
        thrownStack = stack
        pickupStatus = PickupStatus.ALLOWED
    }

    constructor(world: World, shooter: EntityLivingBase, stack: ItemStack) : super(world, shooter) {
        thrownStack = stack
    }

    override fun getArrowStack(): ItemStack {
        return thrownStack.copy()
    }

    override fun findEntityOnPath(start: Vec3d, end: Vec3d): Entity? {
        if (world.isRemote) {
            return null
        }
        var entity: Entity? = null
        val list = world.getEntitiesInAABBexcluding(this, entityBoundingBox.expand(motionX, motionY, motionZ).grow(1.0), ARROW_TARGETS::invoke)
        var d0 = 0.0

        for (entity1 in list) {
            if (entity1 != shootingEntity) {
                val box = entity1.entityBoundingBox.grow(0.30000001192092896)
                val raytraceresult = box.calculateIntercept(start, end)

                if (raytraceresult != null) {
                    val d1 = start.squareDistanceTo(raytraceresult.hitVec)

                    if (d1 < d0 || d0 == 0.0) {
                        entity = entity1
                        d0 = d1
                    }
                }
            }
        }
        return entity
    }

    override fun writeEntityToNBT(compound: NBTTagCompound) {
        super.writeEntityToNBT(compound)
        compound.setTag("ItemTrident", thrownStack.writeToNBT(NBTTagCompound()))
    }

    override fun readEntityFromNBT(compound: NBTTagCompound) {
        super.readEntityFromNBT(compound)
        if (compound.hasKey("ItemTrident")) {
            thrownStack = ItemStack(compound.getCompoundTag("ItemTrident"))
        }
    }

    override fun onHit(raytraceResultIn: RayTraceResult) {
        val entity = raytraceResultIn.entityHit

        if (entity != null) {

            val source: DamageSource = if (shootingEntity !is EntityPlayer) {
                causeTridentDamage(this, this)
            } else {
                causeTridentDamage(this, shootingEntity)
            }

            if (entity.attackEntityFrom(source, getDamageForTrident(entity))) {
                if (entity is EntityLivingBase) {

                    if (shootingEntity is EntityLivingBase) {
                        EnchantmentHelper.applyThornEnchantments(entity, shootingEntity!!)
                        EnchantmentHelper.applyArthropodEnchantments(entity, shootingEntity!!)
                    }

                    arrowHit(entity)

                    if (shootingEntity != null && entity != shootingEntity && entity is EntityPlayer && shootingEntity is EntityPlayer) {
                        (shootingEntity as EntityPlayerMP).connection.sendPacket(SPacketChangeGameState(6, 0.0f))
                    }

                    playSound(Sounds.TRIDENT_PIERCE, 1.0f, 1.0f)

                    motionX *= -0.009999999776482582
                    motionY *= -0.10000000149011612
                    motionZ *= -0.009999999776482582
                    rotationYaw += 180.0f
                    prevRotationYaw += 180.0f

                    if (!hasChanneled && !thrownStack.isEmpty) { // Enchantment handling
                        if (EnchantHelper.hasChanneling(thrownStack)) {
                            if (world.isThundering) {
                                shootingEntity!!.world.addWeatherEffect(EntityLightningBolt(shootingEntity!!.world, posX, posY, posZ, false))
                                playSound(Sounds.TRIDENT_CONDUCTIVIDAD, 5.0f, 1.0f)
                                hasChanneled = true
                            }
                        }
                    }
                }
            } else {
                if (!world.isRemote && motionX * motionX + motionY * motionY + motionZ * motionZ < 0.0010000000474974513) {
                    if (pickupStatus == PickupStatus.ALLOWED) {
                        entityDropItem(arrowStack, 0.1f)
                    }
                    setDead()
                }
            }
        } else {
            motionX = (raytraceResultIn.hitVec.x - posX)
            motionY = (raytraceResultIn.hitVec.y - posY)
            motionZ = (raytraceResultIn.hitVec.z - posZ)
            val f2 = MathHelper.sqrt(motionX * motionX + motionY * motionY + motionZ * motionZ)
            posX -= motionX / f2.toDouble() * 0.05000000074505806
            posY -= motionY / f2.toDouble() * 0.05000000074505806
            posZ -= motionZ / f2.toDouble() * 0.05000000074505806
            inGround = true
            arrowShake = 7

            playSound(Sounds.TRIDENT_IMPACT, 1.0f, 1.0f)

            if (!hasChanneled && EnchantHelper.hasChanneling(thrownStack)) { // Enchantment handling
                val living = if (shootingEntity == null) this else shootingEntity
                living.world.addWeatherEffect(EntityLightningBolt(living.world, posX, posY, posZ, false))
                playSound(Sounds.TRIDENT_CONDUCTIVIDAD, 5.0f, 1.0f)
                hasChanneled = true
            }
        }
    }

    override fun onUpdate() {
        if (inGround) {
            ticksInGround++
        }

        if (shootingEntity is EntityPlayer) {
            val loyalty = EnchantHelper.getLoyalty(thrownStack)

            if (loyalty in 1..3 && !(shootingEntity as EntityPlayer).isSpectator && !shootingEntity!!.isDead && ticksInGround > 100 - loyalty * 20) {
                noClip = true

                if (!isReturning) {
                    playSound(Sounds.TRIDENT_LOYALTY, 1.0f, 1.0f)
                    isReturning = true
                }
                if (ticksInGround >= 100 - loyalty * 20) {
                    val d0 = 0.02 * loyalty.toDouble()
                    val vec3d = Vec3d(shootingEntity!!.posX - posX, shootingEntity!!.posY + shootingEntity!!.eyeHeight.toDouble() - posY, shootingEntity!!.posZ - posZ)
                    motionX += vec3d.x * d0 - motionX * 0.05
                    motionY += vec3d.y * d0 - motionY * 0.05
                    motionZ += vec3d.z * d0 - motionZ * 0.05

                    move(MoverType.SELF, motionX * 0.05, motionY * 0.05, motionZ * 0.05)
                }
            }
        }

        if (world.getClosestPlayerToEntity(this, 1.5) == shootingEntity && isReturning) {
            if (pickupStatus == PickupStatus.ALLOWED) {
                dropTridentStack()
            }
            setDead()
        }

        super.onUpdate()
    }

    override fun getIsCritical(): Boolean {
        return false
    }

    override fun isBurning(): Boolean {
        return false
    }

    override fun setFire(seconds: Int) {
        // Fireproof
    }

    private fun getDamageForTrident(target: Entity): Float {
        val level = EnchantHelper.getImpaling(thrownStack).toFloat()
        var damage = 8f

        if (level > 0 && target.isWet) {
            damage += (level * 1.25).toFloat()
        }
        return damage
    }

    private fun dropTridentStack() {
        thrownStack?.let {
            if (!it.isEmpty) {
                val item = EntityItem(world, posX, posY + 0.1.toFloat().toDouble(), posZ, thrownStack!!)
                item.setNoPickupDelay()

                if (captureDrops) {
                    capturedDrops.add(item)
                } else {
                    world.spawnEntity(item)
                }
            }
        }
    }

    companion object {
        private fun causeTridentDamage(trident: Entity, indirectEntityIn: Entity): DamageSource {
            return EntityDamageSourceIndirect("trident", trident, indirectEntityIn).setProjectile()
        }
    }
}